### PR TITLE
Computationally intensive statistics in Rust

### PR DESCRIPTION
**Pull request by AdilZouitine**
Hello team,

Finally, my contribution is ready; here are the changes.

- The most computationally intensive statistics have been implemented in rust: `Quantile,` `EWMean,` `EWVar,` `IQR,` `Kurtosis,` `PeaktoPeak,` `Skew,` `RollingQuantile,` and `RollingIQR.` The interfacing code is available in the `rust_src` folder
- The GitHub actions set up rust and the wheels build are valid

The stats module of river in rust is [here](https://github.com/online-ml/watermill.rs).

Here are the performance gains in the table below :
```
| Statistics       	| Pure Python (s) 	| Rust binding (s) 	| x times improvement 	|
|------------------	|-----------------	|------------------	|---------------------	|
| quantile         	| 2.359           	| 0.148            	| 15.955              	|
| peak to peak     	| 0.216           	| 0.47             	| 4.609               	|
| EWMean           	| 0.158           	| 0.105            	| 1.512               	|
| EWVar            	| 0.426           	| 0.104            	| 4.075               	|
| IQR              	| 4.541           	| 0.169            	| 26.846              	|
| kurtosis         	| 1.785           	| 0.106            	| 16.872              	|
| skewness         	| 1.086           	| 0.105            	| 10.354              	|
| Rolling Quantile 	| 323.520         	| 77.247           	| 4.573               	|
| Rolling IQR      	| 636.528         	| 76.688           	| 9.113               	|
```
The test is the total time to update 1 million of data. 
For rolling statistics, the window is 1 million and there are 2 million updates.

I also benchmarked the serialization and deserialization time; the order of magnitude is the same: twice as slow to nine times faster.

I had to change the values of the mad test because the first five values of the quantile differ from the python implementation.
The bias in rust is not the same as in python.
But the values remain the same after the first five values.

I had to change some tests for scaling.
I add a new test for pickling stats.

@smastelini  In AdaptiveStandardScaler, the mutable attribute is `alpha`, and in my current implementation is not trivial (but possible) to mutate alpha. I made a comment here https://github.com/online-ml/river/blob/rust_stats/river/preprocessing/scale.py#L553

I made some `hacky` things here: 
https://github.com/online-ml/river/blob/rust_stats/river/stats/quantile.py#L59
https://github.com/online-ml/river/blob/rust_stats/river/stats/quantile.py#L66
https://github.com/online-ml/river/blob/rust_stats/river/stats/quantile.py#L143
https://github.com/online-ml/river/blob/rust_stats/river/stats/iqr.py#L134
https://github.com/online-ml/river/blob/rust_stats/river/stats/iqr.py#L145

This pull request may help to overcome this performance issue: https://github.com/online-ml/river/issues/906

### CODE REVIEW

**Comment by smastelini** [comment]
Hi @AdilZouitine! Good stuff! The rust stuff looks like magic, haha.

I'll take my time to check it, but the benchmarks look good! Regarding the adaptive scaler, I think we can simply remove alpha from the mutable parameters. I changed that just to be able to test adjusting the hyperparameter values of pipelines on the fly. I am sure I can find another pipeline component with mutable parameters to put the algorithm brewing in river-extra to the test.

File: rust_src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,336 @@
+use bincode::{deserialize, serialize};
```
Comment (MaxHalford) [inline rust_src/lib.rs:1]: Is it possible to move the `rust_src` directory somewhere in the `river` directory?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -32,8 +32,11 @@ $ conda create -n river -y python=3.9
 $ conda activate river
 ```
 
+You need a `Rust` compiler you can install it by following this [link](https://www.rust-lang.org/fr/tools/install)
```
Comment (MaxHalford) [inline CONTRIBUTING.md:35]:
```suggestion
You need a Rust compiler which you can install [here](https://www.rust-lang.org/fr/tools/install)
```

File: .github/workflows/pypi.yml [hunk 3]
```diff
@@ -2,25 +2,88 @@ name: pypi
 
 on:
   push:
+    push:
+      branches: ["master" ,"rust_stats"]
```
Comment (MaxHalford) [inline .github/workflows/pypi.yml:6]: To remove

File: .github/workflows/pypi.yml [hunk 4]
```diff
@@ -39,16 +102,16 @@ jobs:
...
-          name: artifact
-          path: dist
+  # upload_pypi:
+  #   needs: [build_wheels, build_sdist]
+  #   runs-on: ubuntu-latest
+  #   steps:
+  #     - uses: actions/download-artifact@v2
+  #       with:
+  #         name: artifact
+  #         path: dist
```
Comment (MaxHalford) [inline .github/workflows/pypi.yml:112]: To uncomment

File: .github/workflows/pypi.yml [hunk 5]
```diff
@@ -39,16 +102,16 @@ jobs:
...
 
-      - uses: pypa/gh-action-pypi-publish@v1.4.2
-        with:
-          user: ${{ secrets.pypi_user }}
-          password: ${{ secrets.pypi_password }}
+  #     - uses: pypa/gh-action-pypi-publish@v1.4.2
+  #       with:
+  #         user: ${{ secrets.pypi_user }}
+  #         password: ${{ secrets.pypi_password }}
```
Comment (MaxHalford) [inline .github/workflows/pypi.yml:117]: To uncomment

File: river/stats/iqr.py [hunk 6]
```diff
@@ -103,32 +113,40 @@ class RollingIQR(stats.base.RollingUnivariate):
...
             return None
-        return q_sup - q_inf
+        return self._rolling_iqr.get()
+
+    @property
+    def window_size(self):
+        return self.window_size_value
+
+    def __repr__(self):
+        # We surcharge this method to avoid this error on rust side:
```
Comment (MaxHalford) [inline river/stats/iqr.py:145]:
```suggestion
        # We override this method to avoid this error on the Rust side:
```

**Review by MaxHalford — APPROVED** [review]

**Pull request merged**